PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/helpers/AzureIconLoader.java

```diff
@@ -0,0 +1,73 @@
+ *
+ * MIT License
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated
+ * documentation files (the "Software"), to deal in the Software without restriction, including without limitation
+ * the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and
+ * to permit persons to whom the Software is furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all copies or substantial portions of
+ * the Software.
+ *
+ * THE SOFTWARE IS PROVIDED *AS IS*, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO
+ * THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
+ * TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+package com.microsoft.intellij.helpers;
+
+import com.google.common.base.Preconditions;
+import com.intellij.icons.AllIcons;
+import com.intellij.openapi.util.IconLoader;
+import com.microsoft.tooling.msservices.serviceexplorer.AzureIconSymbol;
+import org.apache.commons.lang3.StringUtils;
+
+import javax.swing.*;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+public class AzureIconLoader {
+
+    private static final Map<AzureIconSymbol, Icon> ICON_MAP = new ConcurrentHashMap<>();
+
+    static {
+        loadIDEIcons();
+    }
+
+    public static Icon loadIcon(AzureIconSymbol iconSymbol) {
+        Preconditions.checkNotNull(iconSymbol);
+        Preconditions.checkArgument(StringUtils.isNotBlank(iconSymbol.getPath()), "iconPath can not be blank.");
+        Icon icon = ICON_MAP.get(iconSymbol);
+        if (icon == null) {
+            icon = IconLoader.getIcon(iconSymbol.getPath());
+            ICON_MAP.put(iconSymbol, icon);
+        }
+        return icon;
+    }
+
+    private static void loadIDEIcons() {
+        loadIcon(AzureIconSymbol.Common.REFRESH, AllIcons.Actions.Refresh);
+        loadIcon(AzureIconSymbol.Common.CREATE, AllIcons.Welcome.CreateNewProject);
+        loadIcon(AzureIconSymbol.Common.DELETE, AllIcons.Actions.GC);
+        loadIcon(AzureIconSymbol.Common.RESTART, AllIcons.Actions.Restart);
+        loadIcon(AzureIconSymbol.Common.SHOW_PROPERTIES, AllIcons.Actions.Properties);
+    }
+
+    private static Icon loadIcon(AzureIconSymbol iconSymbol, Icon customIcon) {
+        Preconditions.checkNotNull(iconSymbol);
+        Preconditions.checkArgument(StringUtils.isNotBlank(iconSymbol.getPath()), "iconPath can not be blank.");
+        Preconditions.checkNotNull(customIcon);
+        Icon icon = ICON_MAP.get(iconSymbol);
+        if (icon == null) {
+            ICON_MAP.put(iconSymbol, customIcon);
```

Comment:
why loadIcon only if correspond iconSymbon is not registered? I think it may be better to allow developer overwrite existing icon symbol map

Reply:
as a private method, it should be ok. If a same icon symbol is registered twice, it is a bug that developers should resolve it and then to commit.

Reply:
Yes, but developer may not find this issue as there is no errors, an exception may help them find the issue soon.